src/rebar_compiler_dag.erl

```diff
@@ -484,3 +484,63 @@ is_dirty(G) ->
 %% vertices, clear the flag before serializing it.
 clear_dirty(G) ->
     digraph:del_vertex(G, '$r3_dirty_bit').
+
+%% Resolve all the dependencies of a .hrl file transitively. Use a cache to
+%% memoize the resolution.
+resolve_hrl_dependencies(Name, Cache, G) ->
+    case is_hrl_file(Name) of
+        false ->
+            {[Name], Cache};
+        true ->
+            case lookup_hrl(Name, Cache) of
+                {ok, Deps} -> {Deps, Cache};
+                error ->
+                    {Deps, NewCache} = resolve_full_hrl(Name, Cache, G),
+                    {Deps, add_hrl(Name, Deps, NewCache)}
+            end
+    end.
+
+is_hrl_file(Name) ->
+    filename:extension(Name) == ".hrl".
+
+resolve_full_hrl(Name, Cache, G) ->
+    lists:foldl(fun(Dep, {Found, C}) -> {Deps, C1} = resolve_hrl_dependencies(Dep, C, G), {Deps++Found, C1} end,
+                {[], Cache},
+                digraph:out_neighbours(G, Name)).
+
+%% Resolve a file name to an app. Use a cache to
+%% memoize the resolution.
+resolve_file_to_app(File, AppDefs, Cache) ->
+    case lookup_file(File, Cache) of
+        error ->
+            App = resolve_full_file_to_app(File, AppDefs),
+            {App, add_file(File, App, Cache)};
+        {ok, App} ->
+            {App, Cache}
+    end.
+
+resolve_full_file_to_app(_File, []) ->
+    [];
+resolve_full_file_to_app(File, [{App, Dir} | Tail]) ->
+    case lists:prefix(Dir, File) of
+        true ->
+            App;
```

Comment:
would it be possible to use something like `{ok, App} | undefined` to avoid having a return type that just has to be type-checked to work? We can leave the cache as is.

Reply:
I changed the API the whole way through.